### PR TITLE
Add /warpinfo command

### PR DESCRIPTION
**Pull request by Chew**
Get information about a warp.

Currently gets warp location
image

New permissions:
`essentials.warplist` - Access to /warplist command

### CODE REVIEW

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+public class Commandwarpinfo extends EssentialsCommand {
+
+    public Commandwarpinfo() {
+        super("warpinfo");
+    }
+
+    @Override
+    public void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
+        if (args.length == 0) {
+            if (!user.isAuthorized("essentials.warpinfo")) {
```
Comment (Chew) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]: Why is perm only checked if there are no args? Seems like a flaw to me

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            }
+            throw new NotEnoughArgumentsException();
+        }
+        final String warpName = args[0];
+        if (!user.isAuthorized("essentials.warps." + warpName)) {
+            throw new Exception(tl("warpUsePermission"));
+        }
+        final Location warpLocation = ess.getWarps().getWarp(args[0]);
+        showWarpInfoMessage(user.getSource(), warpName, warpLocation);
+    }
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]: see below
```suggestion
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    public void run(final Server server, final CommandSource sender, final String commandLabel, final String[] args) throws Exception {
+        if (args.length == 0) {
+            throw new NotEnoughArgumentsException();
+        }
+        final String warpName = args[0];
+        final Location warpLocation = ess.getWarps().getWarp(args[0]);
+        showWarpInfoMessage(sender, warpName, warpLocation);
+    }
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]: you can reduce code by using the same executor for both console and player as different behavior isn't experienced here.

```suggestion
    @Override
    public void run(final Server server, final CommandSource sender, final String commandLabel, final String[] args) throws Exception {
        if (args.length == 0) {
            throw new NotEnoughArgumentsException();
        }
        if (!sender.isAuthorized("essentials.warpinfo")) {
            throw new Exception(tl("noPerm", "essentials.warpinfo"));
        }
        final String name = args[0];
        final Location loc = ess.getWarps().getWarp(args[0]);
        sender.sendMessage(tl("warpInfo", name));
        sender.sendMessage(tl("whoisLocation", loc.getWorld().getName(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
    }
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            return getAvailableWarpsFor(user);
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    @Override
+    protected List<String> getTabCompleteOptions(final Server server, final CommandSource sender, final String commandLabel, final String[] args) {
+        if (args.length == 1) {
+            return new ArrayList<>(ess.getWarps().getList());
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java:42]: REEEEEEEE STREAMS
```suggestion
            if (ess.getSettings().getPerWarpPermission() && sender.isPlayer()) {
                List<String> list = new ArrayList<>();
                for (String curWarp : ess.getWarps().getList()) {
                    if (sender.isAuthorized("essentials.warps." + curWarp)) {
                        list.add(curWarp);
                    }
                }
                return list;
            }
            return new ArrayList<>(ess.getWarps().getList());
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    protected List<String> getTabCompleteOptions(final Server server, final User user, final String commandLabel, final String[] args) {
+        if (args.length == 1 && user.isAuthorized("essentials.warp.list")) {
+            return getAvailableWarpsFor(user);
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]:
```suggestion
```

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java (hunk 4 above)
Comment (Chew) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java:42]: This was copied from the Commandwarp file :joy:

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java (hunk 5 above)
Comment (Chew) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]: Any reason we're removing the tab complete code?

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java (hunk 1 above)
Comment (mdcfe) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]: The command framework already checks `essentials.warpinfo`, this is redundant

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        sender.sendMessage(tl("whoisLocation", loc.getWorld().getName(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
+    }
+
+    @Override
+    protected List<String> getTabCompleteOptions(final Server server, final CommandSource sender, final String commandLabel, final String[] args) {
+        if (args.length == 1) {
+            if (ess.getSettings().getPerWarpPermission() && sender.isPlayer()) {
+                List<String> list = new ArrayList<>();
+                for (String curWarp : ess.getWarps().getList()) {
+                    if (sender.isAuthorized("essentials.warps." + curWarp)) {
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]:
```suggestion
                    if (sender.isAuthorized("essentials.warps." + curWarp, ess)) {
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public void run(final Server server, final CommandSource sender, final String commandLabel, final String[] args) throws Exception {
+        if (args.length == 0) {
+            throw new NotEnoughArgumentsException();
+        }
+        if (!sender.isAuthorized("essentials.warpinfo")) {
+            throw new Exception(tl("noPerm", "essentials.warpinfo"));
+        }
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]:
```suggestion
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        final String name = args[0];
+        final Location loc = ess.getWarps().getWarp(args[0]);
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]:
```suggestion
        final Location loc = ess.getWarps().getWarp(name);
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        sender.sendMessage(tl("warpInfo", name));
+        sender.sendMessage(tl("whoisLocation", loc.getWorld().getName(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
+    }
+
+    @Override
+    protected List<String> getTabCompleteOptions(final Server server, final CommandSource sender, final String commandLabel, final String[] args) {
+        if (args.length == 1) {
+            if (ess.getSettings().getPerWarpPermission() && sender.isPlayer()) {
+                List<String> list = new ArrayList<>();
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]:
```suggestion
                final List<String> list = new ArrayList<>();
```

**Review by JRoy — CHANGES_REQUESTED** [review]
OOOO

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import com.earth2me.essentials.CommandSource;
+import com.earth2me.essentials.User;
+import net.ess3.api.IUser;
+import org.bukkit.Location;
+import org.bukkit.Server;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]:
```suggestion
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
+package com.earth2me.essentials.commands;
+
+import com.earth2me.essentials.CommandSource;
+import com.earth2me.essentials.User;
+import net.ess3.api.IUser;
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandwarpinfo.java]:
```suggestion
```

**Review by JRoy — DISMISSED** [review]

**Review by JRoy — DISMISSED** [review]

File: Essentials/src/main/resources/messages.properties [hunk 12]
```diff
@@ -932,6 +932,9 @@ walking=walking
 warpCommandDescription=List all warps or warp to the specified location.
 warpCommandUsage=/<command> <pagenumber|warp> [player]
 warpDeleteError=\u00a74Problem deleting the warp file.
+warpInfo=\u00a76Information for warp\u00a7c {0}\u00a76.
```
Comment (mdcfe) [inline Essentials/src/main/resources/messages.properties]:
```suggestion
warpInfo=\u00a76Information for warp\u00a7c {0}\u00a76:
```

**Review by mdcfe — APPROVED** [review]

**Pull request merged**